### ISSUE TITLE
Kubernetes state metrics do not allow configurable instance tags

### ISSUE THREAD

**Issue by mrak** [DataDog/dd-agent#3437]
We saw the notice that agent checks should be reported at the `integrations-core` repository, but are posting this issue here as the file in question that may need modification is a util file in this repository.

Let us know if we should move this issue to integrations-core.

**Output of the [info page](https://help.datadoghq.com/hc/en-us/articles/203764635-Agent-Status-and-Information) **

```
2017-07-13 17:53:31,193 | WARNING | dd.collector | utils.service_discovery.config(config.py:31) | No configuration backend provided for service discovery. Only auto config templates will be used.
====================
Collector (v 5.14.1)
====================

  Status date: 2017-07-13 17:53:20 (11s ago)
  Pid: 24438
  Platform: Linux-4.4.0-1-amd64-x86_64-with
  Python Version: 2.7.12, 64bit
  Logs: <stderr>, /opt/datadog-agent/logs/collector.log

  Clocks
  ======

    NTP offset: 0.0074 s
    System UTC time: 2017-07-13 17:53:31.625029

  Paths
  =====

    conf.d: /opt/datadog-agent/agent/conf.d
    checks.d: /opt/datadog-agent/agent/checks.d

  Hostnames
  =========

    socket-hostname: dd-agent-s2t53
    hostname: kube-worker04.stc.braintreepayments.com
    socket-fqdn: dd-agent-s2t53

  Checks
  ======

    kube_dns (5.14.1:1.0.0)
    -----------------------
      - instance #0 [OK]
      - Collected 121 metrics, 0 events & 0 service checks

    kubernetes_state (5.14.1:1.1.0)
    -------------------------------
      - instance #0 [OK]
      - Collected 988 metrics, 0 events & 68 service checks

    kubernetes (5.14.1:1.1.0)
    -------------------------
      - instance #0 [OK]
      - Collected 148 metrics, 0 events & 3 service checks

    ntp (5.14.1:1.0.0)
    ------------------
      - Collected 0 metrics, 0 events & 0 service checks

    disk (5.14.1:1.0.0)
    -------------------
      - instance #0 [OK]
      - Collected 32 metrics, 0 events & 0 service checks

    docker_daemon (5.14.1:1.2.0)
    ----------------------------
      - instance #0 [OK]
      - Collected 230 metrics, 1 event & 1 service check


  Emitters
  ========

    - http_emitter [OK]
```

**Additional environment details (Operating System, Cloud provider, etc):**

Running in the dd-agent docker container on a Debian host system:

```
$ cat /etc/os-release
PRETTY_NAME="Debian GNU/Linux 8 (jessie)"
NAME="Debian GNU/Linux"
VERSION_ID="8"
VERSION="8 (jessie)"
ID=debian
HOME_URL="http://www.debian.org/"
SUPPORT_URL="http://www.debian.org/support/"
BUG_REPORT_URL="https://bugs.debian.org/"
```

**Steps to reproduce the issue:**
1. Modify an instance in `conf.d/auto_conf/kubernetes_state.yaml` to have tags
    ```yaml
    docker_images:
      - kube-state-metrics

    init_config:
    instances:
      # To enable Kube State metrics you must specify the url exposing the API
      - kube_state_url: http://%%host%%:%%port%%/metrics
        tags:
          - myextratag
    ```
2. Restart collector, forwarder, and dogstatsd through `supervisorctl`
3. Check reported metrics and service checks in the DataDog UI

**Describe the results you received:**

Metrics and service checks reported by the kubernetes state check do not have the tag `myextratag` associated with them.

**Describe the results you expected:**

Metrics and service checks reported by the kubernetes state check to have the tag `myextratag` specified within `kubernetes_state.yaml`.

**Additional information you deem important (e.g. issue happens only occasionally):**

The file in question that would need changes is [utils/kubernetes/kube_state_processor.py](https://github.com/DataDog/dd-agent/blob/master/utils/kubernetes/kube_state_processor.py) as it ignores any tags that are included on the configured instance.

**Comment by hkaj** [comment]
Hi @mrak, thanks for the detailed report.

When you say
> Modify an instance in `conf.d/auto_conf/kubernetes_state.yaml` to have tags

Do you mean in `/conf.d/auto_conf`? If that's the case it's failing as expected. Files in this folder are picked up by the entrypoint [here](https://github.com/DataDog/docker-dd-agent/blob/11.0.5141/entrypoint.sh#L172) and copied to `/etc/dd-agent/conf.d`.

If you want to change the config in a container that's already running you can patch `/etc/dd-agent/conf.d/auto_conf/kubernetes_state.yaml` instead and restart the collector. However I would suggest passing the right template from the start to avoid these manual steps.

**Comment by mrak** [comment]
Thanks for the response!

We should have clarified that we were using `{docker,kubectl} exec` to get into an existing container and were modifying `/opt/datadog-agent/agent/conf.d/auto_conf/kubernetes_state.yaml` directly and subsequently issuing

```
supervisorctl -c /opt/datadog-agent/agent/supervisor.conf restart datadog-agent:forwarder
supervisorctl -c /opt/datadog-agent/agent/supervisor.conf restart datadog-agent:collector
supervisorctl -c /opt/datadog-agent/agent/supervisor.conf restart datadog-agent:dogstatsd
```

If we were to make our change to `kubernetes_state.yaml` permanent for future deployments we would drop the file into `/conf.d/auto_conf` instead.

We are using the Alpine-based `docker-dd-agent`, so the configuration files are under `/opt/datadog-agent/agent` instead of `/etc/dd-agent`.

**Comment by hkaj** [comment]
Thanks for the details.

I just realized this issue will be fixed by 5.15 which is going out tomorrow if all goes well. This version migrates the kube-state-metrics check from KubeStateProcessor (deprecated) to the more generic PrometheusCheck that does handle custom tags: https://github.com/DataDog/dd-agent/blob/5.15.0/checks/prometheus_check.py#L234-L240

Closing but feel free to update again if 5.15 doesn't help (I just tried with the RC and it worked so it should be fine). Thanks again for the precise feedback!
